src/nfa/operations.cc

```diff
@@ -787,3 +787,40 @@ mata::OnTheFlyAlphabet mata::nfa::create_alphabet(const std::vector<Nfa*>& nfas)
 Run mata::nfa::encode_word(const Alphabet* alphabet, const std::vector<std::string>& input) {
     return { .word = alphabet->translate_word(input) };
 }
+
+std::set<mata::Word> mata::nfa::get_words(const Nfa& aut, unsigned max_length) {
+    std::set<mata::Word> result;
+
+    // contains a pair: a state s and the word with which we got to the state s
+    std::vector<std::pair<State, mata::Word>> worklist;
+    // initializing worklist
+    for (State init_state : aut.initial) {
+        worklist.push_back({init_state, {}});
+        if (aut.final.contains(init_state)) {
+            result.insert(mata::Word());
+        }
+    }
+
+    unsigned cur_length = 0;
+    while (!worklist.empty() && cur_length < max_length) {
+        std::vector<std::pair<State, mata::Word>> new_worklist;
```

Comment:
I would optimize working with worklists to have two worklists for the whole time of running the function.

One worklist is the current `worklist` and the second is the current `new_worklist` (both declared in the local scope of the whole function). In each iteration of the while loop, `new_worklist.clear()` is called. And at the end, `worklist.swap(new_worklist)` is called to prevent copy-constructing the worklists again and again (and reusing the previously allocated space).

Setting some initial capacity at the worklists creation could be useful, too.

Reply:
Done